gc/base/MarkingScheme.hpp

```diff
@@ -169,43 +160,49 @@ class MM_MarkingScheme : public MM_BaseVirtual
 
 	void completeScan(MM_EnvironmentBase *env);
 	
-	MMINLINE void
-	scanObject(MM_EnvironmentBase *env, omrobjectptr_t objectPtr, MM_CollectorLanguageInterface::MarkingSchemeScanReason reason)
+	/**
+	 * Public. Called from external context, eg concurrent GC.
+	 */
+	MMINLINE uintptr_t
+	scanObject(MM_EnvironmentBase *env, omrobjectptr_t objectPtr, MM_MarkingDelegate::MarkingSchemeScanReason reason, uintptr_t sizeToDo = UDATA_MAX)
 	{
-		if(MM_CollectorLanguageInterface::SCAN_REASON_PACKET == reason) {
-			env->_markStats._objectsScanned += 1;
+		GC_ObjectScannerState objectScannerState;
+		GC_ObjectScanner *objectScanner = _delegate.getObjectScanner(env, objectPtr, &objectScannerState, reason, &sizeToDo);
+		if (NULL != objectScanner) {
+			bool isLeafSlot = false;
+			GC_SlotObject *slotObject;
+#if defined(J9VM_GC_LEAF_BITS)
+			while (NULL != (slotObject = objectScanner->getNextSlot(isLeafSlot))) {
+#else /* J9VM_GC_LEAF_BITS */
+			while (NULL != (slotObject = objectScanner->getNextSlot())) {
+#endif /* J9VM_GC_LEAF_BITS */
+				/* with concurrentMark mutator may NULL the slot so must fetch and check here */
+				inlineMarkObject(env, slotObject->readReferenceFromSlot(), isLeafSlot);
+			}
 		}
-		uintptr_t sizeScanned = _cli->markingScheme_scanObject(env, objectPtr, reason);
+
 		/* Due to concurrent marking and packet overflow _bytesScanned may be much larger than the total live set
 		 * because objects may be scanned multiple times.
 		 */
-		env->_markStats._bytesScanned += sizeScanned;
-	}
-
-#if defined(OMR_GC_MODRON_CONCURRENT_MARK)
-	MMINLINE uintptr_t
-	scanObjectWithSize(MM_EnvironmentBase *env, omrobjectptr_t objectPtr, MM_CollectorLanguageInterface::MarkingSchemeScanReason reason, UDATA sizeToDo)
-	{
-		if(MM_CollectorLanguageInterface::SCAN_REASON_PACKET == reason) {
+		env->_markStats._bytesScanned += sizeToDo;
```

Comment:
It no longer exists. All sites that previously called this method now call the public version of scanObject(...), which receives size as a parameter. This version updates stats inside, as did previous scanObjectWithSize(). The private version of scanObject(), which is called exclusively from the main completeScan() loop, leaves stats update to caller, avoids NULL object check, and avoids  test of `SCAN_REASON_PACKET == reason` which is necessary in the public version.